### PR TITLE
Move server event filtering logic to rust

### PR DESCRIPTION
**Pull request by devonh**
### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [X] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [X] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: rust/src/events/filter.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * See the GNU Affero General Public License for more details:
+ * <https://www.gnu.org/licenses/agpl-3.0.html>.
+ */
+
+use std::collections::HashMap;
+
+use pyo3::{exceptions::PyValueError, pyfunction, PyResult};
+use ruma_common::OwnedUserId;
+use ruma_events::room::{history_visibility::HistoryVisibility, member::MembershipState};
```
Comment (erikjohnston) [inline rust/src/events/filter.rs:19]: I'm a little bit more sceptical with introducing Ruma into Synapse, it feels like quite a large risk of things getting out of sync between the two projects very easily especially around validation.

Given that the usages we have here are fairly trivial, can we just replace them?

File: synapse/synapse_rust/events.pyi [hunk 2]
```diff
@@ -105,3 +105,13 @@ class EventInternalMetadata:
...
+
+def event_visible_to_server(
+    sender: str,
+    target_server_name: str,
+    history_visibility: str,
+    erased_senders: Mapping[str, bool],
+    partial_state_invisible: bool,
+    memberships: List[Tuple[str, str]],
+) -> bool:
+    """Whether the server is allowed to see the unredacted event"""
```
Comment (erikjohnston) [inline synapse/synapse_rust/events.pyi:117]: We should document the args

File: synapse/visibility.py [hunk 3]
```diff
@@ -726,20 +686,16 @@ def check_event_is_visible(
...
     to_return = []
     for e in events:
-        if include_event_in_output(e):
+        if event_visible_to_server(
+            e.sender,
+            target_server_name,
+            event_to_history_vis[e.event_id],
+            erased_senders,
+            e.event_id in partial_state_invisible_event_ids,
+            list(event_to_memberships.get(e.event_id, {}).values()),
```
[the remark this answers is not in the document]
Comment (devonh) [reply synapse/visibility.py:697]: I'm not sure what you mean here.

Another idea I had would be to change the `memberships` arg to be `StateMap[str]` and then here just pass in `event_to_memberships.get(e.event_id, {})`. This doesn't really change much though.

[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/visibility.py:697]: Sorry, I mean can we not use named arguments, i.e.:

```python
            sender=e.sender,
            target_server_name=target_server_name,
            history_visibility=event_to_history_vis[e.event_id],
            erased_senders=erased_senders,
            partial_state_invisible=e.event_id in partial_state_invisible_event_ids,
            memberships=list(event_to_memberships.get(e.event_id, {}).values()),
```

[the remark this answers is not in the document]
Comment (devonh) [reply synapse/visibility.py:697]: ah I see now. Yeah no problem :)
(that makes way more sense haha)

**Pull request merged**